appscale/tools/node_layout.py

```diff
@@ -367,202 +317,259 @@ def is_valid_simple_format(self):
-    elif len(master_nodes) > 1:
-      return self.invalid("Only one master is allowed")
+    # Dont need the hash any more, make a nodes list
+    nodes = node_hash.values()
 
-    master_node = master_nodes[0]
+    self.distribute_unassigned_roles(nodes, role_count, master_node)
 
-    login_nodes = []
-    for node in nodes:
-      if node.is_role('login'):
-        login_nodes.append(node)
+    if self.infrastructure in InfrastructureAgentFactory.VALID_AGENTS:
+      if not self.min_vms:
+        self.min_vms = len(nodes)
+      if not self.max_vms:
+        self.max_vms = len(nodes)
 
-    # If a login node was not specified, make the master into the login node
-    if not login_nodes:
-      master_node.add_role('login')
+    rep = self.is_database_replication_valid(nodes)
+    if not rep:
+      return rep
 
-    # by this point, somebody has a login role, so now's the time to see if we
-    # need to override their ip address with --login_host
-    if self.login_host is not None:
+    self.nodes = nodes
+
+    return True
+
+  def is_valid_advanced_format(self):
+    """Checks to see if this NodeLayout represents an acceptable (new) advanced
+    deployment strategy, and if so, constructs self.nodes from it.
+    
+    Returns:
+      True if the deployment strategy is valid.
+    Raises:
+      BadConfigurationException with reason if the deployment strategy is not 
+        valid.
+    """
+    # Keep track of whether the deployment is valid while going through.
+    node_hash = {}
+    role_count = {
+      'appengine': 0,
+      'master': 0,
+      'memcache': 0,
+      'taskqueue': 0,
+      'zookeeper': 0,
+      'login': 0
+    }
+    node_count = 1
+    db_master_created = False
+    tq_master_created = False
+    master_node = None
+    login_found = False
+    # Loop through the list of "node sets", which are grouped by role.
+    for node_set in self.input_yaml:
+      # If the key nodes is mapped to an integer it should be a cloud
+      # deployment.
+      cloud = True if isinstance(node_set.get('nodes'), int) else False
+
+      # In cloud deployments, set the fake public ips to node-#.
+      if cloud:
+        ips_list = ["node-{}".format(node_count + i) \
+                    for i in xrange(node_set.get('nodes'))]
+      # Otherwise get the ips and validate them.
+      else:
+        ip_or_ips = node_set.get('nodes')
+        ips_list = ip_or_ips if isinstance(ip_or_ips, list) else [ip_or_ips]
+        # Validate that the ips_list are either node-id or ip addresses.
+        for ip in ips_list:
+          self.is_cloud_ip(ip)
+
+      if len(ips_list) == 0:
+        self.invalid("Node amount cannot be zero.")
+
+      # Get the roles.
+      role_or_roles = node_set.get('roles')
+      roles = role_or_roles if isinstance(role_or_roles, list) else \
+        [role_or_roles]
+
+      # Immediately fail if we have more than one node for master.
+      if 'master' in roles and master_node and len(ips_list) > 1:
```

Comment:
The first part of the condition should be enough for that, right? The extra condition just prevents the exception from being thrown if the size of the current ips_list is greater than 1. For example, if master was already defined in a different node_set and it is defined again in this set and the size of the current ips_list is greater than 1, the master_node will just get overwritten.